### PR TITLE
Yet more tests

### PR DESCRIPTION
**Pull request by jerith**
Now that we finally have most of the test infrastructure set up, we can write more tests. :-)

### CODE REVIEW

File: spec/integration/integration_spec.rb [hunk 1]
```diff
@@ -16,15 +16,12 @@
...
+      describe 'gluster_peer create' do
         it 'should add a single peer (manifest)' do
           expect(@fake_gluster.peer_hosts).to eq([])
-          x = apply_manifest(<<-'END'
-            node default {
-              gluster_peer { 'gfs1.local': }
-            }
-            END
-          )
+          x = apply_node_manifest(<<-'MANIFEST')
```
Comment (JayH5) [inline spec/integration/integration_spec.rb]: Wow this syntax.. where does the closing `)` even belong? :open_mouth:

File: spec/unit/helpers.rb [hunk 2]
```diff
@@ -6,3 +6,11 @@ def props(providers)
     providers.map { |p| props(p) }
   end
 end
+
+def res_hash(resources)
+  Hash[resources.map { |r| [r.name, r] }]
+end
+
+def res_providers(resources)
+  resources.map { |r| r.provider && r.provider.name }
```
Comment (JayH5) [inline spec/unit/helpers.rb]: `&&`? What is that doing?

File: spec/integration/integration_spec.rb (hunk 1 above)
Comment (jerith) [inline spec/integration/integration_spec.rb]: The content of the heredoc only starts on the line after the starting syntax, and that syntax is replaced by the content. Thus, you can treat the `<<-'END'` as if it were a string literal containing the content. (For example, you can do `foo = <<-'END'.gsub(/re/, repl)` followed by the lines of content.) The end token must be the only non-whitespace on its line, so you can't do `END )` or whatever.

I usually find that sort of thing hard to read, but in this case the `MANIFEST` serves as an end-of-content indicator and the extra line for the closing paren just wastes vertical space.

File: spec/unit/helpers.rb (hunk 2 above)
Comment (jerith) [inline spec/unit/helpers.rb]: `&&` is a logical AND operator. (Ruby also has `and`, which I think differs only in precedence.) `foo && bar` means "if `foo` is falsey (`false` or `nil`, basically) return `foo`, otherwise return `bar`". (Python's logical operators have very similar behaviour.) I'm using it here to avoid explicit checks for `nil` because those turn a one-liner into three lines of variable juggling.

Comment (JayH5) [inline spec/unit/helpers.rb]: Ah, it's making sense now, thanks.

**Comment by jerith** [comment]
Ready for review.

There are still lots of edge cases that need tests, but most of them also need the correct behaviour to be implemented. That can happen in other branches.

File: spec/helpers/fake_gluster.rb [hunk 3]
```diff
@@ -178,6 +235,15 @@ def add_volumes(*volumes)
     volumes.each { |volume| add_volume(volume) }
   end
 
+  def get_volume(name)
+    (volume,) = @volumes.select { |vol| vol.name == name }
```
Comment (JayH5) [inline spec/helpers/fake_gluster.rb]: We only care about the first volume? Every time I see one of these "filter/select and then take the first one" I ask something like this... I know it's usually some way to avoid nil values or something.. but it's not immediately clear. Why not use `find`?

File: spec/helpers/fake_gluster.rb [hunk 4]
```diff
@@ -185,48 +251,103 @@ def gluster(*args)
...
+          :opErrno => 30800,
+          :opErrstr => "Host #{peer} is not in 'Peer in Cluster' state",
+      )) unless all_peers.include? peer
+    end
+    @volumes << volume
+    format_doc(volume.short_xml(make_cli_elem('volCreate')))
+  end
+
+  def cmd_volume_delete(name)
+    (volume,) = @volumes.select { |vol| vol.name == name }
```
Comment (JayH5) [inline spec/helpers/fake_gluster.rb]: Can we not use `get_volume` here?

File: lib/puppet/provider/gluster_volume/gluster_volume.rb [hunk 5]
```diff
@@ -4,7 +4,7 @@
   :gluster_volume, :parent => Puppet::Provider::Gluster) do
 
   has_command(:gluster, 'gluster') do
-    environment :HOME => "/tmp"
+    environment :HOME => '/tmp'
```
Comment (JayH5) [inline lib/puppet/provider/gluster_volume/gluster_volume.rb]: Can you do a grep for `"` across all the source so we can get all the `"` -> `'` once and for all?

File: spec/helpers/fake_gluster.rb (hunk 3 above)
Comment (jerith) [inline spec/helpers/fake_gluster.rb]: Bleh, I was thinking in Pythonland where `(foo,) = somelist` explodes unless the list contains exactly one element. I want to make sure there's exactly one result, so I guess I'll need to write it more verbosely.

I moved away from `Enumerable#find` because that returns the first result even if there are several matches.

File: spec/helpers/fake_gluster.rb (hunk 4 above)
Comment (jerith) [inline spec/helpers/fake_gluster.rb]: We can, thanks. (This code predates `get_volume`.)

File: lib/puppet/provider/gluster_volume/gluster_volume.rb (hunk 5 above)
Comment (jerith) [inline lib/puppet/provider/gluster_volume/gluster_volume.rb]: This should be the last file. We still need `""` for places where we're evaluating expressions inside the string, though.

Comment (JayH5) [inline lib/puppet/provider/gluster_volume/gluster_volume.rb]: Cool.

**Comment by JayH5** [comment]
`x`? `trx`? What do these variable names mean? (I hadn't noticed the use of `x` before)

**Comment by jerith** [comment]
`x` is a placeholder. `trx` is a transaction object containing the results of applying the catalog.

**Comment by JayH5** [comment]
:+1:
